### PR TITLE
Feature/failed charges

### PR DESCRIPTION
**Pull request by johantjellden**

### CODE REVIEW

File: src/main/java/com/hedvig/notificationService/customerio/dto/ChargeFailedEvent.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import java.time.LocalDate
+import javax.validation.constraints.PositiveOrZero
+
+data class ChargeFailedEvent(
+  @JsonProperty(required = true)
+  val terminationDate: LocalDate?,
+  @JsonProperty(required = true)
+  @get:PositiveOrZero
```
Comment (FredrikAreschoug) [inline src/main/java/com/hedvig/notificationService/customerio/dto/ChargeFailedEvent.kt:11]: Is there a PositiveOrOne? and what happens if its null?

File: src/test/java/com/hedvig/notificationService/customerio/web/OnChargeFailedEventTriggerCustomerioEventTest.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.hedvig.notificationService.service.FirebaseNotificationService
+import io.mockk.mockk
+import io.mockk.slot
+import io.mockk.verify
+import org.junit.jupiter.api.Test
+
+class OnChargeFailedEventTriggerCustomerioEventTest {
+
+    @Test
+    internal fun `first test`() {
```
Comment (FredrikAreschoug) [inline src/test/java/com/hedvig/notificationService/customerio/web/OnChargeFailedEventTriggerCustomerioEventTest.kt:21]: Yeah maybe we should add more and rename this test at some point

File: src/main/java/com/hedvig/notificationService/customerio/EventHandler.kt [hunk 3]
```diff
@@ -38,4 +44,20 @@ class EventHandler(
             repo.save(state)
         }
     }
+
+    fun onFailedChargeEvent(memberId: String, chargeFailedEvent: ChargeFailedEvent) {
+        clients[Workspace.SWEDEN]?.sendEvent(memberId, chargeFailedEvent.toMap(memberId))
```
Comment (johantjellden) [inline src/main/java/com/hedvig/notificationService/customerio/EventHandler.kt:49]: We just have to get the workspace for the current member before we make this call.

File: src/main/java/com/hedvig/notificationService/customerio/dto/ChargeFailedEvent.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    val chargeFailedReason: ChargeFailedReason
+) {
+
+    fun toMap(memberId: String) = mapOf(
+        "name" to "ChargeFailedEvent",
+        "member_id" to memberId,
+        "number_of_failed_charges" to numberOfFailedCharges,
+        "charges_left_before_termination" to chargesLeftBeforeTermination,
+        "termination_date" to terminationDate,
+        "charge_failed_reason" to chargeFailedReason
```
Comment (johantjellden) [inline src/main/java/com/hedvig/notificationService/customerio/dto/ChargeFailedEvent.kt:26]: These ones need to live in the "data" part of the event:
```
{
"name": "ChargeFailedEvent",
"data": {
    "member": "memberId",
    . . .
}
}
```

File: src/test/java/com/hedvig/notificationService/customerio/web/OnChargeFailedEventTriggerCustomerioEventTest.kt (hunk 2 above)
Comment (johantjellden) [reply src/test/java/com/hedvig/notificationService/customerio/web/OnChargeFailedEventTriggerCustomerioEventTest.kt:21]: Yes"

**Pull request merged**